[ADD] VoIP: Axivox Documentation Grouping (#5256)
odoo/documentation: content/applications/general/voip/axivox/dial_plan_advanced.rst

==================
Dial plan advanced
==================

Companies have a lot of incoming calls every day, but many do not want their teams to answer calls
24 hours a day, 7 days a week. Using *Axivox* advanced dial plan features, the process can be
automated and routing can be setup for all scenarios. This way customers are never left waiting or
frustrated because they can not get in touch with anyone.

Using the advanced elements in dial plans, companies can automate call routing for certain days or
times, like company holidays. Companies can also allow callers to enter extensions themselves and
get transferred automatically using a digital receptionist, so an administrative team does not have
to be available around the clock. There is even the option to route callers depending on where they
are calling from in the world, to maximize efficiency.

.. seealso::

>    For more information on basic dial plans and adding elements visit :doc:`dial_plan_basics`.

Author: I've added in an important admonition block.

Advanced elements
=================

In Axivox dial plans there are two advanced elements that can be used.

- :guilabel:`Record`: recording feature is enabled (requires plan change, enabled in Settings).
- :guilabel:`Caller ID`: replace the caller ID by the called number or free text.

> To add one of these elements navigate to :guilabel:`Dial plans` in the left menu of the `Axivox
> management console <https://manage.axivox.com>`_. The :guilabel:`Record` element records calls that
> are routed through this element and requires an additional plan change in Axivox.

Reviewer: ```suggestion
To add one of these elements, navigate to the :guilabel:`Dial plans` page, located in the menu on
the left side of the `Axivox management console <https://manage.axivox.com>`_.

.. important::
   The :guilabel:`Record` element records calls that are routed through this element, and requires
   an additional plan change in Axivox.
```

> To enable recording on Axivox, navigate to :guilabel:`Settings` in the `Axivox management console
> <https://manage.axivox.com>`_. Then go to the :guilabel:`Recording` drop-down menu near the bottom
> of the page. Select :guilabel:`Enabled` to enable recording using the :guilabel:`Record` element in
> a dial plan.

Reviewer: ```suggestion
To enable recording on Axivox, navigate to :guilabel:`Settings` in the `Axivox management console
<https://manage.axivox.com>`_. Then, go to the :guilabel:`Recording` drop-down menu, near the bottom
of the page. From there, select :guilabel:`Enabled` from the drop-down menu to enable recording
using the :guilabel:`Record` element in a dial plan.
```

Reviewer: It might be smart to explain why this Recording drop-down menu might not be editable/changeable in the Settings page, as it's currently grayed-out and unable to change.

> The :guilabel:`Caller ID` element allows for the replacement of the caller ID downstream, after
> routing. Upon adding the :guilabel:`Caller ID` element and double clicking it to configure it, two
> options appear. The first is a :guilabel:`Free text` field where any text can be input to replace
> the caller ID. The second option is :guilabel:`Replace the caller ID by the called number`. This
> option replaces the caller's ID with the :guilabel:`Incoming number`.

Reviewer: ```suggestion
The :guilabel:`Caller ID` element allows for the replacement of the caller ID downstream, after
routing. Upon adding the :guilabel:`Caller ID` element, and double-clicking it to configure it, two
options appear.

The first is a :guilabel:`Free text` field, where any text can be input to replace the caller ID.
The second option is :guilabel:`Replace the caller ID by the called number`. This option replaces
the caller's ID with the :guilabel:`Incoming number`.
```
I'm actually not seeing any Caller ID field option on the General Settings page on Axivox...Can you confirm where this field is located?

Author: Noted, this is added in.

.. important::
   A company may want to use the :guilabel:`Caller ID` element to replace the :guilabel:`Incoming

>    number` so that employees or outside transfers cannot see the number and information is kept
>    private.

Reviewer: ```suggestion
   number`, so employees, or outside transfers, cannot see the number, and information is kept
   private.
```

Basic routing elements
======================

> Basic routing elements in Axivox dial plans provide extension based routing. This can be done by
> adding either a :guilabel:`Menu` to numerically link the dial-by-numbers to an action, or by using
> a :guilabel:`Digital Receptionist` to automatically route or listen for an extension based on a key
> input from the caller. The main difference between the two elements is that the :guilabel:`Digital
> Receptionist` does **not** need to be pre-configured numerically with actions. Instead, it acts as a
> virtual receptionist.

Reviewer: ```suggestion
Basic routing elements in Axivox dial plans provide extension-based routing. This can be done by
adding either a *Menu* to numerically link the dial-by-numbers to an action, or by using a *Digital
Receptionist* to automatically route, or listen for an extension based on a key
input from the caller.

The main difference between the two elements is that the Digital Receptionist does **not** need to
be pre-configured numerically with actions. Instead, it acts as a virtual receptionist.
```

Reviewer: Where can any of these features/elements be accessed? I'm not seeing this information/option(s)...

Author: I've added a basic process into each section

Author: Good call on explaining the menu feature a bit better

- :guilabel:`Menu`: add a dial-by-number directory and configured downstream actions (not
  terminal).

> - :guilabel:`Digital Receptionist`: attached a virtual dispatcher to listen for extensions to
>   connect to.

Reviewer: ```suggestion
- :guilabel:`Digital Receptionist`: attach a virtual dispatcher to listen for extensions.
```

   element that plays an :guilabel:`Audio message` that reads, 'Enter the extension of the person
   you are trying to reach'. Following the first :guilabel:`Play a file` element, add the
   :guilabel:`Digital receptionist` element, followed by another :guilabel:`Play a file` element
   which states, 'That is not a valid extension'. This last element is in place to close the loop
   should the caller not enter a correct extension. Finally, this last :guilabel:`Play a file`
   element is looped back into the :guilabel:`Menu` element.

Advanced routing elements
=========================

Advanced routing elements route calls automatically as they are received into the incoming
number(s). This can be configured using geo-location, whitelisting or time-based variables. Calls
pass through a filter prior to their final destination and are routed based on the set variable(s).

- :guilabel:`Dispatcher`: create a call filter to route traffic based on geo-location of the caller
  ID.
- :guilabel:`Access list`: create a tailored access list with VIP customer preference.
- :guilabel:`Time condition`: create time conditions to route incoming traffic around holidays or

>   other sensitive time-frames.

Author: I've added in a admonition block

click :guilabel:`Save`. Then, on the :guilabel:`Access list` element in the dial plan, three paths
or routes are available to link to further actions. Unknown calls can be routed through the regular
menu flow by adding a :guilabel:`Menu` element and connecting it to :guilabel:`Unknown`.
:guilabel:`Refused` calls can be routed to the :guilabel:`Hang up` element. Lastly
:guilabel:`Authorized` callers can be sent to a specific extension or queue.

.. image:: dial_plan_advanced/access-list.png
   :align: center
   :alt: Access list element highlighted in an example dial plan.

Switches
========

A *switch* in Axivox is a simple activated/deactivated route action. These can be activated or
chosen quickly, allowing for quick routing changes without altering the dial plan. Alternate routes
can be set to jump to, at a moments notice, based on variables such as availability or to adjust
traffic flow for any number of reasons. Axivox allows for a simple On/Off switch and a multi-switch,
which can have several paths to choose from.

- :guilabel:`Switch`: a manual on/off control that can divert traffic based on whether it is opened
  (On) or closed (off).

> - :guilabel:`Multi-Switch`: a mechanism to create paths and turn them on and off to divert incoming
>   calls.
> 
> Basic switch
> ------------
> 
> A :guilabel:`Switch` can be set in the `Axivox management console <https://manage.axivox.com>`_ by
> navigating to :guilabel:`Switches` in the left menu. To create a new switch click :guilabel:`Add a
> switch` and then configure a :guilabel:`Name` for it.
> 
> Toggle the :guilabel:`Switch` :guilabel:`On` and :guilabel:`Off` from the :guilabel:`State` column
> in the Switches dashboard (click :guilabel:`Switches` in the left menu).
> 
> This :guilabel:`On` / :guilabel:`Off` state automatically routes traffic in a dial plan in which
> this switch is set. The traffic travels to the :guilabel:`Active` route when :guilabel:`On` is
> toggled in the switch. The call traffic travels to the :guilabel:`Inactive` route when
> :guilabel:`Off` is toggled in the switch. Changes can be made on the fly, just be sure to click
> :guilabel:`Apply changes` to implement the them.
> 
> Add a switch to dial plan
> ~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> To add a :guilabel:`Switch` to a dial plan, navigate to `Axivox management console
> <https://manage.axivox.com>`_ and click on :guilabel:`Dial plans` in the left menu. Then select or
> create a dial plan, and click on the drop-down menu. Select :guilabel:`Switch` and then click
> :guilabel:`Add`. Double click on the element to set the :guilabel:`Switch`.
> 
> .. image:: dial_plan_advanced/switch.png
>    :align: center
>    :alt: Switch configuration in a dial plan, with inactive and active routes highlighted.
> 
> Multi-switch
> ------------
> 
> A :guilabel:`Multi-switch` in Axivox is a switch where multiple paths can be configured and switched
> between. To configure and set a :guilabel:`Multi-switch`, navigate to `Axivox management console
> <https://manage.axivox.com>`_. Then, click on the :guilabel:`Switches` menu item in the left menu.
> Toggle to the :guilabel:`Multi-switch` tab to create or set a pre-configured
> :guilabel:`Multi-switch`.
> 
> To create a new :guilabel:`Multi-switch` click :guilabel:`Create new`. Enter a :guilabel:`Name` for
> the :guilabel:`Multi-switch` and then enter the :guilabel:`Available choice`. Enter one
> :guilabel:`Available choice` per line. Do **not** duplicate any entries.
> 
> To select the :guilabel:`State` of the :guilabel:`Multi-switch` click the drop-down menu next to the
> :guilabel:`Multi-switch` name. Whichever :guilabel:`State` is chosen is the route that is followed
> in the dial plan. The :guilabel:`State` can be edited on the fly, just be sure to :guilabel:`Apply
> changes`.
> 
> Add a multi-switch to dial plan
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 
> To add a :guilabel:`Multi-switch` to a dial plan, first, navigate to `Axivox management console
> <https://manage.axivox.com>`_ and click on :guilabel:`Dial plans` in the left menu. Then select or
> create a dial plan, and click on the drop-down menu. Select :guilabel:`Multi-switch` and then click
> :guilabel:`Add`. Double click on the element to set the :guilabel:`Switch`.
> 
> .. image:: dial_plan_advanced/multi-switch.png
>    :align: center
>    :alt: Multi-switch configuration in a dial plan, with chosen route highlighted.

Reviewer: ```suggestion
Access list scenario
--------------------

An :guilabel:`Access List` element in a dial plan allows for the routing of certain numbers, and
disallows (denies) other numbers.

After adding an :guilabel:`Access List` element to a dial plan, it can be configured by
double-clicking on the element directly in the :guilabel:`Dialplan Editor` pop-up window.

Two fields appear where regular expressions can based in the :guilabel:`Allow` and :guilabel:`Deny`
fields of the :guilabel:`Access List` pop-up window.

.. example::
   For a very important customer, their number can be set in the :guilabel:`Allow` field, and these
   callers can be sent directly to management.

.. tip::
   A regular expression (shortened to regex or regexp), sometimes also referred to as a rational
   expression, is a sequence of characters that specifies a match pattern in text.

.. image:: dial_plan_advanced/access-config.png
   :align: center
   :alt: Access list element configuration with the allow/deny fields highlighted.

.. example::
   - `2\\d\\d`: validates numbers from 200 to 299
   - `02\\d*`: validates all numbers beginning with 02
   - `0017165551212`: validates the number (0017165551212)

After setting the :guilabel:`Allow` and :guilabel:`Deny` fields with regular expressions or numbers,
click :guilabel:`Save` on the :guilabel:`Access List` pop-up window.

Then, on the :guilabel:`Access list` element in the dial plan, three paths (or routes) are available
to link to further actions.

Unknown calls can be routed through the regular menu flow by adding a :guilabel:`Menu` element, and
connecting it to the :guilabel:`Unknown` path. :guilabel:`Refused` calls can be routed to the
:guilabel:`Hang up` element. Lastly, :guilabel:`Authorized` callers can be sent to a specific
extension or queue.

.. image:: dial_plan_advanced/access-list.png
   :align: center
   :alt: Access list element highlighted in an example dial plan.

Switches
========

A :guilabel:`Switch` element in Axivox is a simple activated/deactivated route action.

These can be activated or chosen quickly, allowing for quick routing changes, without altering the
dial plan.

Alternate routes can be set to jump to, at a moment's notice, based on variables, such as
availability, or to adjust traffic flow for any number of reasons.

Axivox allows for a simple on/off switch, and a multi-switch, which can have several paths to choose
from.

- :guilabel:`Switch`: a manual on/off control that can divert traffic, based on whether it is opened
  (On) or closed (Off).
- :guilabel:`Multi-Switch`: a mechanism to create paths, and turn them on and off, to divert
  incoming calls.

Basic switch
------------

A :guilabel:`Switch` can be set in the `Axivox management console <https://manage.axivox.com>`_ by
navigating to :guilabel:`Switches` in the left menu. To create a new switch click :guilabel:`Add a
switch`, configure a :guilabel:`Name` for it, and click :guilabel:`Save`.

Then, toggle the desired switch to either :guilabel:`On` or :guilabel:`Off`, from the
:guilabel:`State` column in the :guilabel:`Switches`` dashboard (accessible via the
:guilabel:`Switches` option in the left menu).

This :guilabel:`On` / :guilabel:`Off` state automatically routes traffic in a dial plan, in which
this switch is set.

The traffic travels to the :guilabel:`Active` route when :guilabel:`On` is toggled in the switch.
The call traffic travels to the :guilabel:`Inactive` route when :guilabel:`Off` is toggled in the
switch.

Changes can be made on the fly, just be sure to click :guilabel:`Apply changes` to implement the
them.

Add a switch to dial plan
~~~~~~~~~~~~~~~~~~~~~~~~~

To add a :guilabel:`Switch` to a dial plan, navigate to `Axivox management console
<https://manage.axivox.com>`_, and click on :guilabel:`Dial plans` in the left menu. Then, click
:guilabel:`Visual Editor` next to the desired dial plan to open the :guilabel:`Dialplan Editor`
pop-up window. 

Then, from the :guilabel:`New element` drop-down menu, select :guilabel:`Switch`, and then click
:guilabel:`Add`. Double-click on the element to further configure the :guilabel:`Switch` element.

.. image:: dial_plan_advanced/switch.png
   :align: center
   :alt: Switch configuration in a dial plan, with inactive and active routes highlighted.

Multi-switch
------------

A :guilabel:`Multi-Switch` element in Axivox is a switch where multiple paths can be configured, and
switched between.

To configure, and set, a :guilabel:`Multi-Switch` element, navigate to `Axivox management console
<https://manage.axivox.com>`_. Then, click on the :guilabel:`Switches` menu item in the left menu.

Toggle to the :guilabel:`Multi-switch` tab to create, or set, a pre-configured
:guilabel:`Multi-Switch` element.

To create a new :guilabel:`Multi-Switch`, click :guilabel:`Create new`. Then, enter a
:guilabel:`Name` for the element, and then enter the :guilabel:`Available choice`. Enter one
:guilabel:`Available choice` per line. Do **not** duplicate any entries.

Remember to click :guilabel:`Save` when done.

To select the :guilabel:`State` of the :guilabel:`Multi-switch`, click the drop-down menu next to
the :guilabel:`Multi-switch` name, under the :guilabel:`Multi-switch` tab on the
:guilabel:`Switches` page.

The :guilabel:`State` chosen is the route that is followed in the dial plan. The :guilabel:`State`
can be edited on the fly, just be sure to click :guilabel:`Apply changes`.

Add a multi-switch to dial plan
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To add a :guilabel:`Multi-Switch` element to a dial plan, navigate to `Axivox management console
<https://manage.axivox.com>`_, and click :guilabel:`Dial plans` in the left menu.

Then, select or create a dial plan. Next, click :guilabel:`Visual Editor` on the desired dial plan.

On the :guilabel:`Dialplan Editor` pop-up window that appears, click on the :guilabel:`New element`
drop-down menu, and select :guilabel:`Multi-Switch`. Then, click :guilabel:`Add`.

Double-click on the element to further configure the :guilabel:`Switch` element.

.. image:: dial_plan_advanced/multi-switch.png
   :align: center
   :alt: Multi-switch configuration in a dial plan, with chosen route highlighted.
```

option replaces the caller's ID with the :guilabel:`Incoming number`.

.. important::
   A company may want to use the :guilabel:`Caller ID` element to replace the :guilabel:`Incoming
   number` so that employees or outside transfers cannot see the number and information is kept
   private.

Basic routing elements
======================

Basic routing elements in Axivox dial plans provide extension based routing. This can be done by
adding either a :guilabel:`Menu` to numerically link the dial-by-numbers to an action or by using
a :guilabel:`Digital Receptionist` to automatically route or listen for an extension based on a key
input from the caller. The main difference between the two elements is that the :guilabel:`Digital
Receptionist` does not need to be pre-configured numerically with actions. Instead it acts as a
virtual receptionist.

- :guilabel:`Menu` - Add a dial-by-number directory and configured downstream actions (not
  terminal).

- :guilabel:`Digital Receptionist` - Attached a virtual dispatcher to listen for extensions to
  connect to.

Digital receptionist scenario
-----------------------------

> The :guilabel:`Digital receptionist` element is a listen-feature that accurately routes callers

Author: No it isn't.

.. image:: dial_plan_advanced/access-config.png
   :align: center
   :alt: Access list element configuration with the allow/deny fields highlighted.

.. example::
   - `2\\d\\d`: validates numbers from 200 to 299
   - `02\\d*`: validates all numbers beginning with 02
   - `0017165551212`: validates the number (0017165551212)

After setting the :guilabel:`Allow` and :guilabel:`Deny` fields with regular expressions or numbers,
click :guilabel:`Save`. Then, on the :guilabel:`Access list` element in the dial plan, three paths
or routes are available to link to further actions. Unknown calls can be routed through the regular
menu flow by adding a :guilabel:`Menu` element and connecting it to :guilabel:`Unknown`.
:guilabel:`Refused` calls can be routed to the :guilabel:`Hang up` element.

.. image:: dial_plan_advanced/access-list.png
   :align: center
   :alt: Access list element highlighted in an example dial plan.

Switches
========

A switch in Axivox is a simple activated / deactivated route action. These can be activated or
chosen quickly, allowing for quick routing changes without altering the dial plan. Alternate routes

> can be set to jump to, in a moments notice based on variables such as availability or to adjust

Reviewer 2: ```suggestion
can be set to jump to at a moments notice, based on variables such as availability or to adjust
```

which can have several paths to choose from.

- :guilabel:`Switch` - A switch is a manually on/off control that can divert traffic based on
  whether it is opened (On) or closed (off).

- :guilabel:`Multi-Switch` - A mechanism to create paths and turn them on and off to divert incoming
  calls.

Basic switch
------------

A :guilabel:`Switch` can be set in the `Axivox management console <https://manage.axivox.com>`_ by
navigating to :guilabel:`Switches` in the left menu. To create a new switch click :guilabel:`Add a
switch` and then configure a :guilabel:`Name` for it.

Toggle the :guilabel:`Switch` :guilabel:`On` and :guilabel:`Off` from the :guilabel:`State` column
in the Switches dashboard (click :guilabel:`Switches` in the left menu).

This :guilabel:`On` / :guilabel:`Off` state automatically routes traffic in a dial plan in which
this switch is set. The traffic travels to the :guilabel:`Active` route when :guilabel:`On` is
toggled in the switch. The call traffic travels to the :guilabel:`Inactivate` route when
:guilabel:`Off` is toggled in the switch. Changes can be made on the fly, just be sure to

> click :guilabel:`Apply changes` to implement the changes.

Reviewer 2: Early line break, move 'click' from line 248 to 247

Add a switch to dial plan
~~~~~~~~~~~~~~~~~~~~~~~~~

> To add a :guilabel:`Switch` to a dial plan, first, navigate to `Axivox management console

Reviewer 2: ```suggestion
To add a :guilabel:`Switch` to a dial plan, navigate to `Axivox management console
```

<https://manage.axivox.com>`_ and click on :guilabel:`Dial plans` in the left menu. Then select or
create a dial plan, and click on the drop-down menu. Select :guilabel:`Switch` and then click
:guilabel:`Add`. Double click on the element to set the :guilabel:`Switch`.

.. image:: dial_plan_advanced/switch.png
   :align: center
   :alt: Switch configuration in a dial plan, with inactive and active routes highlighted.

Multi-switch
------------

A :guilabel:`Multi-switch` in Axivox is a switch where multiple paths can be configured and switched

> between. To configure and set a :guilabel:`Multi-switch` first, navigate to `Axivox management

Reviewer 2: ```suggestion
between. To configure and set a :guilabel:`Multi-switch`, navigate to `Axivox management
```

console <https://manage.axivox.com>`_. Then, click on the :guilabel:`Switches` menu item in the left
menu. Toggle to the :guilabel:`Multi-switch` tab to create or set a pre-configured
:guilabel:`Multi-switch`.

To create a new :guilabel:`Multi-switch` click :guilabel:`Create new`. Enter a :guilabel:`Name` for
the :guilabel:`Multi-switch` and then enter the :guilabel:`Available choice`(s). Enter one

> :guilabel:`Available choice` per line and do not duplicate any.

Reviewer 2: ```suggestion
:guilabel:`Available choice` per line. Do **not** duplicate any entries.
```

To select the :guilabel:`State` of the :guilabel:`Multi-switch` click the drop-down menu next to the
:guilabel:`Multi-switch` name. Whichever :guilabel:`State` is chosen is the route that is followed
in the dial plan. The :guilabel:`State` can be edited on the fly, just be sure to :guilabel:`Apply
changes`.

Add a multi-switch to dial plan
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To add a :guilabel:`Multi-switch` to a dial plan, first, navigate to `Axivox management

Reviewer 2: Early line break, move console from line 284 to 283

their customers to certain queues, or users based on the customer's geo-location.

To specify all numbers behind a certain country code, or area code, include `\d+` after the country
code, or country code + area code.

.. image:: dial_plan_advanced/dispatcher.png
   :align: center
   :alt: Dispatcher configuration panel, with name, regular expression and add a line highlighted.

.. example::
  - `02\\d+`: validates the numbers starting with 02
  - `00\\d+`: validates all numbers beginning with 00
  - `0052\\d+` validates all numbers beginning with 0052 (Mexico country code)
  - `001716\\d+`: validates all numbers beginning with 001716 (USA country code + Western New York
    area code)

.. tip::
   A regular expression (shortened to regex or regexp), sometimes also referred to as a rational
   expression, is a sequence of characters that specifies a match pattern in text. In other words, a
   match is made within the given range of numbers.

When the desired configurations are complete on the :guilabel:`Dispatch` pop-up window, be sure to
click :guilabel:`Save`.

Upon doing so, the :guilabel:`Dispatcher` element appears with different routes available to

> configure, based on the :guilabel:`Regular Expressions` that were set.

Author: Noted, got them all

   - `02\\d*`: validates all numbers beginning with 02
   - `0017165551212`: validates the number (0017165551212)

After setting the :guilabel:`Allow` and :guilabel:`Deny` fields with regular expressions or numbers,
click :guilabel:`Save` on the :guilabel:`Access List` pop-up window.

Then, on the :guilabel:`Access list` element in the dial plan, three paths (or routes) are available
to link to further actions.

Unknown calls can be routed through the regular menu flow by adding a :guilabel:`Menu` element, and
connecting it to the :guilabel:`Unknown` path. :guilabel:`Refused` calls can be routed to the
:guilabel:`Hang up` element. Lastly, :guilabel:`Authorized` callers can be sent to a specific
extension or queue.

.. image:: dial_plan_advanced/access-list.png
   :align: center
   :alt: Access list element highlighted in an example dial plan.

Switches
========

A :guilabel:`Switch` element in Axivox is a simple activated/deactivated route action.

These can be activated or chosen quickly, allowing for quick routing changes, without altering the
dial plan.

Alternate routes can be set to jump to, at a moment's notice, based on variables, such as

> availability, or to adjust traffic flow for any number of reasons.

Author: Alternate routes can be configured, so that in a moments notice, they can be switched to. This could
be for new availability, or to adjust traffic flow for any number of reasons.
